pilot/pkg/serviceregistry/kube/controller/endpointslice.go

```diff
@@ -102,50 +102,48 @@ func (esc *endpointSliceController) onEventInternal(_, ep *v1.EndpointSlice, eve
 	} else {
 		esc.updateEndpointSlice(ep)
 	}
+
 	hostnames := esc.c.hostNamesForNamespacedName(namespacedName)
 	// Trigger EDS push for all hostnames.
 	esc.pushEDS(hostnames, namespacedName.Namespace)
 
+	// Now check if we need to do a full push for the service.
+	// If the service is headless, we need to do a full push if service exposes TCP ports
+	// to create IP based listeners. For pure HTTP headless services, we only need to push NDS.
 	name := serviceNameForEndpointSlice(esLabels)
 	namespace := ep.GetNamespace()
 	svc := esc.c.services.Get(name, namespace)
 	if svc == nil || svc.Spec.ClusterIP != corev1.ClusterIPNone || svc.Spec.Type == corev1.ServiceTypeExternalName {
 		return
 	}
 
-	configs := []types.NamespacedName{}
-	pureHTTP := true
+	configsUpdated := sets.New[model.ConfigKey]()
+	supportsOnlyHTTP := true
 	for _, modelSvc := range esc.c.servicesForNamespacedName(config.NamespacedName(svc)) {
 		// skip push if it is not exported
 		if modelSvc.Attributes.ExportTo.Contains(visibility.None) {
 			continue
 		}
 
-		configs = append(configs, types.NamespacedName{Name: modelSvc.Hostname.String(), Namespace: svc.Namespace})
-
 		for _, p := range modelSvc.Ports {
 			if !p.Protocol.IsHTTP() {
-				pureHTTP = false
+				supportsOnlyHTTP = false
```

Comment:
isn't this changing the behavior?

Before: check for all services to decide `pureHTTP`
After: we update pureHTTP for each service, but check it after each one

I guess maybe its not likely to be a big change since servicesForNamespacedName should return similar services?